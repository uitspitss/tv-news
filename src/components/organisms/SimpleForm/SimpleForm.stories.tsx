import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import base from 'paths.macro';

import { SimpleForm, SimpleFormProps } from './SimpleForm';

const path = base.replace(
  /^.+\/(atoms|molecules|organisms|templates|pages)\/(.+)\//,
  '$1/$2',
);

export default {
  title: `${path}`,
  component: SimpleForm,
  argTypes: {
    isOpen: { control: 'boolean' },
  },
} as Meta;

const Template: Story<SimpleFormProps> = (args) => <SimpleForm {...args} />;

export const Desktop = Template.bind({});

export const Mobile = Template.bind({});
Mobile.parameters = {
  viewport: {
    defaultViewport: 'iphonex',
  },
};
