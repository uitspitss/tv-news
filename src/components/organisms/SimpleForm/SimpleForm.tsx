import React, { memo, VFC } from 'react';
import tw, { styled } from 'twin.macro';
import { useForm, UseFormMethods, Controller } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import * as z from 'zod';

import { TextInput } from '~/components/atoms/TextInput';
import { Button } from '~/components/atoms/buttons/Button';
import { Select } from '~/components/atoms/Select';
import { Checkbox } from '~/components/atoms/Checkbox';
import { SimpleSelect } from '~/components/atoms/SimpleSelect';

export type SimpleFormProps = {
  /**
   * class for twin.macro
   */
  className?: string;
  /**
   * onValid func
   */
  onValid: (data: FormValues) => void;
};

type Props = {
  methods: UseFormMethods<FormValues>;
  onValid: (data: FormValues) => void;
} & SimpleFormProps;

const schema = z.object({
  email: z.string().email(),
  sex: z.union([
    z.string().min(1, { message: 'have to select' }),
    z.object({ value: z.string(), label: z.string() }),
  ]),
  agreement: z.boolean().refine((val) => val, { message: 'have to check' }),
});

export type FormValues = z.infer<typeof schema>;

const Component: VFC<Props> = (props) => {
  const { className, methods, onValid } = props;
  const { register, handleSubmit, errors, control } = methods;

  return (
    <div className={className}>
      <div className="headerText">
        <h1>Form</h1>
      </div>
      <form onSubmit={handleSubmit(onValid)}>
        <div className="inputFields">
          <div>
            <div className="email">
              <TextInput
                name="email"
                type="email"
                label="email"
                placeholder="Email"
                ref={register}
                error={errors.email?.message}
              />
            </div>
            <div className="Sex">
              <Controller
                name="sex"
                control={control}
                defaultValue=""
                render={(renderProps) => (
                  <Select
                    label="sex"
                    placeholder="Sex"
                    options={[
                      { value: 'male', label: 'Male' },
                      { value: 'female', label: 'Female' },
                      { value: 'others', label: 'Others' },
                    ]}
                    error={errors.sex?.message}
                    {...renderProps}
                  />
                )}
              />
            </div>
            <div className="linkNews">
              <SimpleSelect
                name="like-news"
                label="Which news do you like?"
                options={[
                  { value: '', label: 'choose news type' },
                  { value: 'tech', label: 'Tech' },
                  { value: 'political', label: 'Political' },
                  { value: 'peace', label: 'Peace' },
                ]}
                error={errors.sex?.message}
              />
            </div>
            <div className="Check">
              <Checkbox
                name="agreement"
                label="agreement"
                ref={register}
                error={errors.agreement?.message}
              />
            </div>
            <div className="submitButton">
              <Button type="submit" primary size="large">
                Submit
              </Button>
            </div>
          </div>
        </div>
      </form>
    </div>
  );
};

const StyledComponent = styled(Component)`
  .headerText {
    ${tw`flex flex-col text-center mb-4`}
    & > h1 {
      ${tw`sm:text-3xl text-2xl font-medium mb-4 text-primary-500`}
    }

    & > p {
      ${tw`lg:w-2/3 mx-auto leading-relaxed text-base`}
    }
  }

  .inputFields {
    ${tw`lg:w-1/2 md:w-2/3 mx-auto`}
    & > div {
      ${tw`flex flex-col justify-center`}
      & > div {
        ${tw`p-2`}
        &.submitButton {
          ${tw`mx-auto`}
        }
      }
    }
  }

  .error-text {
    ${tw`text-secondary-500`}
  }
`;

/**
 * Simple Form
 */
export const SimpleForm = memo((props: SimpleFormProps) => {
  const methods = useForm<FormValues>({
    resolver: zodResolver(schema),
  });

  return (
    <StyledComponent
      {...props}
      {...{
        methods,
      }}
    />
  );
});
