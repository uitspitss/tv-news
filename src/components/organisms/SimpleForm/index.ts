export { SimpleForm } from './SimpleForm';
export type { FormValues } from './SimpleForm';
