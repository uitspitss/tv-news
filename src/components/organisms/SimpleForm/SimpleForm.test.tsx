import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import selectEvent from 'react-select-event';

import { SimpleForm } from './SimpleForm';

describe('SimpleForm', () => {
  test('snapshot test', () => {
    const onValid = () => {};
    const { container } = render(<SimpleForm onValid={onValid} />);
    expect(container).toMatchSnapshot();
  });

  describe('Behavior', () => {
    const submitFn = jest.fn().mockImplementation((data) => data);

    beforeEach(() => {
      jest.clearAllMocks();
      render(<SimpleForm onValid={submitFn} />);
    });

    test('it should display required error when value is invalid', async () => {
      userEvent.click(screen.getByRole('button'));

      expect(await screen.findAllByRole('alert')).toHaveLength(4);
    });

    test('it should display required error when email is invalid', async () => {
      userEvent.type(screen.getByLabelText('email'), 'test');

      fireEvent.click(screen.getByRole('button'));

      expect(screen.getByLabelText('email')).toHaveValue('test');

      expect(await screen.findAllByRole('alert')).toHaveLength(4);
    });

    test('it should display required error when message is invalid', async () => {
      userEvent.type(screen.getByLabelText('email'), 'mail@example.com');

      fireEvent.click(screen.getByRole('button'));

      expect(await screen.findAllByRole('alert')).toHaveLength(3);
      expect(submitFn).not.toBeCalled();
      expect(await screen.findByLabelText('email')).toHaveValue(
        'mail@example.com',
      );
    });

    test('it should display required error when checkbox is invalid', async () => {
      userEvent.type(screen.getByLabelText('email'), 'mail@example.com');
      await selectEvent.select(screen.getByLabelText('sex'), 'Male');

      fireEvent.click(screen.getByRole('button'));

      expect(await screen.findByRole('alert')).toBeInTheDocument();
      expect(submitFn).not.toBeCalled();
      expect(await screen.findByLabelText('email')).toHaveValue(
        'mail@example.com',
      );
    });

    test('it should not display errors when values are valid', async () => {
      userEvent.type(screen.getByLabelText('email'), 'mail@example.com');
      await selectEvent.select(screen.getByLabelText('sex'), 'Male');
      fireEvent.click(screen.getByLabelText('agreement'));

      fireEvent.click(screen.getByRole('button'));

      expect(screen.queryByRole('alert')).not.toBeInTheDocument();
      expect(await screen.findByLabelText('email')).toHaveValue(
        'mail@example.com',
      );
      expect(await screen.findByLabelText('agreement')).toBeChecked();

      await waitFor(() => {
        expect(submitFn).toBeCalledTimes(1);
        expect(submitFn.mock.results[0].value).toEqual({
          email: 'mail@example.com',
          sex: { value: 'male', label: 'Male' },
          agreement: true,
        });
      });
    });
  });
});
