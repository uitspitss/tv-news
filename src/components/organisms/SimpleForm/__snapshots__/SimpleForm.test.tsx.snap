// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleForm snapshot test 1`] = `
.emotion-0 .headerText {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  text-align: center;
  margin-bottom: 1rem;
}

.emotion-0 .headerText>h1 {
  font-size: 1.5rem;
  line-height: 2rem;
  font-weight: 500;
  margin-bottom: 1rem;
  --tw-text-opacity: 1;
  color: rgba(107, 114, 128, var(--tw-text-opacity));
}

@media (min-width: 640px) {
  .emotion-0 .headerText>h1 {
    font-size: 1.875rem;
    line-height: 2.25rem;
  }
}

.emotion-0 .headerText>p {
  margin-left: auto;
  margin-right: auto;
  line-height: 1.5rem;
  font-size: 1rem;
}

@media (min-width: 1024px) {
  .emotion-0 .headerText>p {
    width: 66.666667%;
  }
}

.emotion-0 .inputFields {
  margin-left: auto;
  margin-right: auto;
}

@media (min-width: 768px) {
  .emotion-0 .inputFields {
    width: 66.666667%;
  }
}

@media (min-width: 1024px) {
  .emotion-0 .inputFields {
    width: 50%;
  }
}

.emotion-0 .inputFields>div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-0 .inputFields>div>div {
  padding: 0.5rem;
}

.emotion-0 .inputFields>div>div.submitButton {
  margin-left: auto;
  margin-right: auto;
}

.emotion-0 .error-text {
  --tw-text-opacity: 1;
  color: rgba(99, 102, 241, var(--tw-text-opacity));
}

.emotion-2>input {
  width: 100%;
  --tw-bg-opacity: 1;
  background-color: rgba(243, 244, 246, var(--tw-bg-opacity));
  border-radius: 0.25rem;
  border-width: 1px;
  --tw-border-opacity: 1;
  border-color: rgba(156, 163, 175, var(--tw-border-opacity));
}

.emotion-2>input:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
  --tw-border-opacity: 1;
  border-color: rgba(99, 102, 241, var(--tw-border-opacity));
}

.emotion-2>.error {
  --tw-text-opacity: 1;
  color: rgba(239, 68, 68, var(--tw-text-opacity));
}

.emotion-4 .select__control {
  --tw-bg-opacity: 1;
  background-color: rgba(243, 244, 246, var(--tw-bg-opacity));
  border-radius: 0.25rem;
  border-width: 1px;
  --tw-border-opacity: 1;
  border-color: rgba(156, 163, 175, var(--tw-border-opacity));
}

.emotion-4 .select__control:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
  --tw-border-opacity: 1;
  border-color: rgba(99, 102, 241, var(--tw-border-opacity));
}

.emotion-4 .select__value-container {
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
}

.emotion-4 .error {
  --tw-text-opacity: 1;
  color: rgba(239, 68, 68, var(--tw-text-opacity));
}

.emotion-6 {
  position: relative;
  box-sizing: border-box;
}

.emotion-7 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: hsl(0, 0%, 100%);
  border-color: hsl(0, 0%, 80%);
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  cursor: default;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  min-height: 38px;
  outline: 0!important;
  position: relative;
  -webkit-transition: all 100ms;
  transition: all 100ms;
  box-sizing: border-box;
}

.emotion-7:hover {
  border-color: hsl(0, 0%, 70%);
}

.emotion-8 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 2px 8px;
  -webkit-overflow-scrolling: touch;
  position: relative;
  overflow: hidden;
  box-sizing: border-box;
}

.emotion-9 {
  color: hsl(0, 0%, 50%);
  margin-left: 2px;
  margin-right: 2px;
  position: absolute;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -moz-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  box-sizing: border-box;
}

.emotion-10 {
  margin: 2px;
  padding-bottom: 2px;
  padding-top: 2px;
  visibility: visible;
  color: hsl(0, 0%, 20%);
  box-sizing: border-box;
}

.emotion-11 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  box-sizing: border-box;
}

.emotion-12 {
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  background-color: hsl(0, 0%, 80%);
  margin-bottom: 8px;
  margin-top: 8px;
  width: 1px;
  box-sizing: border-box;
}

.emotion-13 {
  color: hsl(0, 0%, 80%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 8px;
  -webkit-transition: color 150ms;
  transition: color 150ms;
  box-sizing: border-box;
}

.emotion-13:hover {
  color: hsl(0, 0%, 60%);
}

.emotion-14 {
  display: inline-block;
  fill: currentColor;
  line-height: 1;
  stroke: currentColor;
  stroke-width: 0;
}

.emotion-15>select {
  width: 100%;
  --tw-bg-opacity: 1;
  background-color: rgba(243, 244, 246, var(--tw-bg-opacity));
  border-radius: 0.25rem;
  border-width: 1px;
  --tw-border-opacity: 1;
  border-color: rgba(156, 163, 175, var(--tw-border-opacity));
}

.emotion-15>select:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
  --tw-border-opacity: 1;
  border-color: rgba(99, 102, 241, var(--tw-border-opacity));
}

.emotion-15>.error {
  --tw-text-opacity: 1;
  color: rgba(239, 68, 68, var(--tw-text-opacity));
}

.emotion-17> {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-17>label>input {
  margin-right: 0.5rem;
  border-radius: 0.25rem;
}

.emotion-17>div>.error {
  --tw-text-opacity: 1;
  color: rgba(239, 68, 68, var(--tw-text-opacity));
}

.emotion-19 {
  font-size: 1rem;
  line-height: 1.5rem;
  border-radius: 0.25rem;
  border-style: none;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
  --tw-bg-opacity: 1;
  background-color: rgba(107, 114, 128, var(--tw-bg-opacity));
  --tw-text-opacity: 1;
  color: rgba(229, 231, 235, var(--tw-text-opacity));
  font-size: 1.125rem;
  line-height: 1.75rem;
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <div
      class="headerText"
    >
      <h1>
        Form
      </h1>
    </div>
    <form>
      <div
        class="inputFields"
      >
        <div>
          <div
            class="email"
          >
            <div
              class="emotion-2 emotion-3"
            >
              <label
                for="email"
              >
                email
              </label>
              <input
                aria-invalid="false"
                aria-label="email"
                id="email"
                name="email"
                placeholder="Email"
                type="email"
              />
            </div>
          </div>
          <div
            class="Sex"
          >
            <div
              class="emotion-4 emotion-5"
            >
              <label
                for="sex"
              >
                sex
              </label>
              <div
                class=" emotion-6"
                id="sex"
              >
                <div
                  class="select__control emotion-7"
                >
                  <div
                    class="select__value-container emotion-8"
                  >
                    <div
                      class="select__placeholder emotion-9"
                    >
                      Sex
                    </div>
                    <div
                      class="emotion-10"
                    >
                      <div
                        class="select__input"
                        style="display: inline-block;"
                      >
                        <input
                          aria-autocomplete="list"
                          aria-label="sex"
                          autocapitalize="none"
                          autocomplete="off"
                          autocorrect="off"
                          id="react-select-2-input"
                          spellcheck="false"
                          style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                          tabindex="0"
                          type="text"
                          value=""
                        />
                        <div
                          style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="select__indicators emotion-11"
                  >
                    <span
                      class="select__indicator-separator emotion-12"
                    />
                    <div
                      aria-hidden="true"
                      class="select__indicator select__dropdown-indicator emotion-13"
                    >
                      <svg
                        aria-hidden="true"
                        class="emotion-14"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <input
                  name="sex"
                  type="hidden"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="linkNews"
          >
            <div
              class="emotion-15 emotion-16"
            >
              <label
                for="like-news"
              >
                Which news do you like?
              </label>
              <select
                aria-invalid="false"
                aria-label="Which news do you like?"
                id="like-news"
                name="like-news"
              >
                <option
                  value=""
                >
                  choose news type
                </option>
                <option
                  value="tech"
                >
                  Tech
                </option>
                <option
                  value="political"
                >
                  Political
                </option>
                <option
                  value="peace"
                >
                  Peace
                </option>
              </select>
            </div>
          </div>
          <div
            class="Check"
          >
            <div
              class="emotion-17 emotion-18"
            >
              <label
                for="agreement"
              >
                <input
                  aria-invalid="false"
                  aria-label="agreement"
                  id="agreement"
                  name="agreement"
                  type="checkbox"
                />
                <span
                  class="label"
                >
                  agreement
                </span>
              </label>
            </div>
          </div>
          <div
            class="submitButton"
          >
            <button
              class="emotion-19 emotion-20"
              type="submit"
            >
              Submit
            </button>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
